// ader/webpack.config.js

const path = require('path');

module.exports = {
    mode: 'development',
    entry: {
        'index': path.join(__dirname, './src/frame/index.js'),
        'react-division': path.join(__dirname, './src/core/index.js'),
        'react-mlc': path.join(__dirname, './src/core2/index.js')
        // index: path.join(__dirname, './demo/index.js')
    },
    output: {
        path: path.join(__dirname, './lib'),
        // path: path.join(__dirname, './target'),
        filename: '[name].js',
        library: '[name]',
        libraryTarget: 'umd',
        chunkFilename: '[name].js'
    },
    devtool: 'cheap-module-source-map',
    externals: {
        'react': 'react',
        'react-dom': 'react-dom'
    },
    module: {
        rules: [{
            test: /\.(jsx|js)$/,
            use: {
                loader: 'babel-loader',
                options: {
                    presets: ['@babel/env', '@babel/react'],
                    plugins: [
                        'transform-class-properties'
                    ]
                }
            }
        }, {
            test: /\.less/,
            use: {
                loader: 'less-loader'
            }
        }]
    }
};
